pkg/controller/daemon_controller.go

```diff
@@ -0,0 +1,432 @@
+			glog.Errorf("Couldn't get object from tombstone %+v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*api.Pod)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not a pod %+v", obj)
+			return
+		}
+	}
+	if dc := dm.getPodDaemonController(pod); dc != nil {
+		dcKey, err := controllerKeyFunc(dc)
+		if err != nil {
+			glog.Errorf("Couldn't get key for object %+v: %v", dc, err)
+		}
+		dm.expectations.DeletionObserved(dcKey)
+		dm.enqueueController(dc)
+	}
+}
+
+func (dm *DaemonManager) addNode(obj interface{}) {
+	glog.Infoln("Processing Node")
+	dm.enqueueAllControllers()
+}
+
+func (dm *DaemonManager) updateNode(old, cur interface{}) {
+	oldNode := old.(*api.Node)
+	curNode := cur.(*api.Node)
+	if api.Semantic.DeepEqual(oldNode.Name, curNode.Name) && api.Semantic.DeepEqual(oldNode.Namespace, curNode.Namespace) && api.Semantic.DeepEqual(oldNode.Labels, curNode.Labels) {
+		// A periodic relist will send update events for all known pods.
+		return
+	}
+	glog.Infoln("Updating node")
+	dm.enqueueAllControllers()
+}
+
+func (dm *DaemonManager) getNodesToDaemonPods(dc *api.DaemonController) (map[string][]string, error) {
+	nodeToDaemonPods := make(map[string][]string) // nodeToDaemonPods["A"] stores names of daemon pods running on node "A"
+	daemonPods, err := dm.podStore.Pods(dc.Namespace).List(labels.Set(dc.Spec.Selector).AsSelector())
+	if err != nil {
+		return nodeToDaemonPods, err
+	}
+	for i := range daemonPods.Items {
+		nodeName := daemonPods.Items[i].Spec.NodeName
+		nodeToDaemonPods[nodeName] = append(nodeToDaemonPods[nodeName], daemonPods.Items[i].Name)
+	}
+	return nodeToDaemonPods, nil
+}
+
+func (dm *DaemonManager) manageDaemons(dc *api.DaemonController) {
+	// Find out which nodes are running the daemon pods selected by dc.
+	nodeToDaemonPods, err := dm.getNodesToDaemonPods(dc)
+	if err != nil {
+		glog.Errorf("Error getting node to daemon pod mapping for daemon controller %+v: %v", dc, err)
+	}
+
+	// For each node, if the node is running the daemon pod but isn't supposed to, kill the daemon
+	// pod. If the node is supposed to run the daemon, but isn't, create the daemon on the node.
+	nodeList, err := dm.nodeStore.List()
+	if err != nil {
+		glog.Errorf("Couldn't get list of nodes when adding daemon controller %+v: %v", dc, err)
+	}
+	var nodesNeedingDaemons, podsToDelete []string
+	for i := range nodeList.Items {
+		nodeName := nodeList.Items[i].Name
+		nodeSelector := labels.Set(dc.Spec.Template.Spec.NodeSelector).AsSelector()
+		shouldRun := nodeSelector.Matches(labels.Set(nodeList.Items[i].Labels))
+		daemonPodNames, isRunning := nodeToDaemonPods[nodeName]
+		if shouldRun && !isRunning {
+			// If daemon pod is supposed to be running on node, but isn't, create daemon pod.
+			nodesNeedingDaemons = append(nodesNeedingDaemons, nodeName)
+		} else if shouldRun && len(daemonPodNames) > 1 {
+			// If daemon pod is supposed to be running on node, but more than 1 daemon pod is running, delete the excess daemon pods.
```

Comment:
This can either be a TODO or an actual fix, but you should sort these pods by creation time so that you preserve the oldest. (I don't actually care what the order is as long as it is deterministic, but this is what we will do for RC.) You'll probably want to make your list be of type *api.Pod for this instead of just keeping the name.